ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1966,6 +1966,59 @@ extension SignalProducerType {
 	}
 }
 
+extension SignalProducerType {
+	
+	/// Creates the new signal producer that forwards all `Next` events from input and restarts the original one each time the input signal producer sends `Completed` event. `Completed` event is sent for `Error` event from the input signal. Each restart should be delayed for delayInterval on the given scheduler.
+	///
+	/// - parameters:
+	///   - delayInterval: Interval to delay each restart loop
+	///   - scheduler: A scheduler to deliver restarted sequence on
+	///
+	/// - returns: A producer that which applies the criteria above
+	public func restart(withDelay delayInterval: NSTimeInterval, on scheduler: DateSchedulerType) -> SignalProducer<Value, Error> {
+		
+		return SignalProducer { observer, outerDisposable in
+			
+			func recursiveRestartWithDelay() {
+				self.delay(delayInterval, onScheduler: scheduler)
+					.startWithSignal { signal, innerDisposable in
+						setupObservations(signal, innerDisposable: innerDisposable)
+				}
+			}
+			
+			func setupObservations(signal: ReactiveCocoa.Signal<Value, Error>, innerDisposable: Disposable) {
+				outerDisposable += innerDisposable
+				
+				signal.observe { event in
+					
+					switch event {
+						
+					case .Next(let value):
+						observer.sendNext(value)
+						
+					case .Completed:
+						recursiveRestartWithDelay()
+						
+					case .Failed:
+						observer.sendCompleted()
+					default: break
```

Comment:
As a side note, from our experience we find it safer to avoid having `default` cases in switches and to handle all cases explicitly. This will make the compiler point us to all places of interest when a new case is added to the enum.